<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [2]:
import pandas as pd
import psycopg2

In [12]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [206]:
query_3_1 = f'''
SELECT count(id)
FROM vacancies
'''

In [207]:
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [208]:
query_3_2 = f'''
SELECT count(id)
FROM employers
'''

In [209]:
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\3588252056.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [210]:
query_3_3 = f'''
SELECT count(id)
FROM areas
'''

In [211]:
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\2018224393.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [212]:
query_3_4 = f'''
SELECT count(id)
FROM industries
'''

In [213]:
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\2172479966.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

Представлена обширная выборка данных с количеством резюме в почти 50000. Осталось понять, что из себя представляют данные и насколько они соответствуют целям проекта.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [214]:
query_4_1 = f'''
SELECT 
    a.name,
    COUNT(area_id) cnt
FROM vacancies v
JOIN areas a ON a.id = v.area_id
GROUP BY a.id
ORDER BY 2 DESC
'''

In [215]:
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\347208320.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


name   cnt
0                      Москва  5333
1             Санкт-Петербург  2851
2                       Минск  2112
3                 Новосибирск  2006
4                      Алматы  1892
..                        ...   ...
764                    Кизляр     1
765                    Джизак     1
766                    Эртиль     1
767                  Арсеньев     1
768  Тура (Красноярский край)     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [216]:
query_4_2 = f'''
SELECT COUNT(id)
FROM vacancies
WHERE (salary_from IS NOT NULL) OR (salary_to IS NOT NULL)
'''

In [217]:
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\2168503709.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [218]:
query_4_3 = f'''
SELECT 
    AVG(salary_from) avg_salary_from,
    AVG(salary_to) avg_salary_to
FROM vacancies
WHERE (salary_from IS NOT NULL) OR (salary_to IS NOT NULL)
'''

In [219]:
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\2222372203.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0     71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [220]:
query_4_4 = f'''
SELECT 
    schedule,
    employment,
    COUNT(id)
FROM vacancies
GROUP BY schedule, employment
ORDER BY COUNT(id) DESC
'''

In [221]:
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\2981602351.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [1]:
query_4_5 = f'''
SELECT experience, COUNT(id)
FROM vacancies
GROUP BY experience
ORDER BY COUNT(id) ASC
'''

In [5]:
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_7720\2601166264.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

У половины вакансий заполнено хотя бы одно из двух полей с зарплатой, что хорошо для создания нашей модели.

Основная часть вакансий нацелена на полную занятость. Это логично

На Москву + Санкт-Петербург приходится 1/6 всех вакансий

Средняя вилка от 70 до 110 тыс. руб. 

В большинстве своём у вакансий указан необходимый опыт работы от 1 года до 3 лет. Можно предположить, что компаниям нужны люди который уже более-менее разбираются в том, чего от них ожидают. При этом кандидаты с большим опытом могут потребовать бОльшую з/п, что не каждая компания готова предложить.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [92]:
query_5_1 = f'''
with x as(
SELECT 
    e.name,
    COUNT(v.id) cnt
FROM vacancies v
JOIN employers e ON e.id = v.employer_id
GROUP BY e.name
ORDER BY COUNT(v.id) DESC)

-- выводим первое место по количеству вакансий
(SELECT *
FROM x
LIMIT 1)

-- объединяем с 5 местом по количеству вакансий
UNION
(SELECT *
FROM x
LIMIT 1 OFFSET 4)
ORDER BY 2 DESC
'''

In [93]:
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_7720\1792814825.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name   cnt
0         Яндекс  1933
1  Газпром нефть   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [226]:
query_5_2 = f'''
-- запрос для вывода количества работадателей по районам
WITH ae AS(
SELECT
    a.name,
    COUNT(e.name) employers_count
FROM areas a
LEFT JOIN employers e ON a.id = e.area
GROUP BY a.name
),
-- запрос для вывода количества вакансий по районам
av AS(
SELECT
    a.name,
    COUNT(v.id) vacancies_count
FROM areas a
LEFT JOIN vacancies v ON a.id = v.area_id
GROUP BY a.name
)

-- объединение двух запросов по регоинам
SELECT
    ae.name,
    ae.employers_count,
    av.vacancies_count
FROM ae
JOIN av ON ae.name = av.name
ORDER BY 3 ASC, 2 DESC
'''


In [227]:
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\3341481329.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  employers_count  vacancies_count
0                 Россия              410                0
1              Казахстан              207                0
2     Московская область               75                0
3     Краснодарский край               19                0
4               Беларусь               18                0
...                  ...              ...              ...
1357              Алматы              721             1892
1358         Новосибирск              573             2006
1359               Минск             1115             2112
1360     Санкт-Петербург             2217             2851
1361              Москва             5864             5333

[1362 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [228]:
query_5_3 = f'''
SELECT
    e.name,
    COUNT(DISTINCT v.area_id) areas_count
FROM vacancies v
JOIN employers e ON v.employer_id = e.id
GROUP BY e.id
ORDER BY 2 DESC
'''

In [229]:
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\2398953733.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  areas_count
0                                      Яндекс          181
1                                  Ростелеком          152
2                                  Спецремонт          116
3                      Поляков Денис Иванович           88
4                                   ООО ЕФИН            71
...                                       ...          ...
14901                            НПП Авиатрон            1
14902              Центр дистанционных торгов            1
14903  Городские Телекоммуникационные Системы            1
14904                       Введенский, Отель            1
14905              Группа Компаний Автоимпорт            1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [230]:
query_5_4 = f'''
-- Из общего числа работадателей вычитаем количество работадателей, у которых указана сфера деятельности
SELECT COUNT(DISTINCT e.id) - COUNT(DISTINCT ea.employer_id) emp_without_ind
FROM employers e
LEFT JOIN employers_industries ea ON e.id = ea.employer_id
'''

In [231]:
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\3114411492.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


emp_without_ind
0             8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [232]:
query_5_5 = f'''
SELECT
    e.name,
    COUNT(ea.industry_id) industry_count
FROM employers e
JOIN employers_industries ea ON e.id = ea.employer_id
GROUP BY e.name
HAVING COUNT(ea.industry_id) = 4
ORDER BY 1
LIMIT 1 OFFSET 2
'''

In [233]:
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\1650419958.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  industry_count
0  2ГИС               4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [234]:
query_5_6 = f'''
SELECT
    COUNT(ea.employer_id)
FROM employers_industries ea
JOIN industries i ON ea.industry_id = i.id
WHERE i.name = 'Разработка программного обеспечения'
'''

In [235]:
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\1169039133.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [8]:
import requests # Импортируем библиотеку requests
from bs4 import BeautifulSoup # Импортируем библиотеку BeautifulSoup

url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России' # Определяем значение URL страницы для запроса
response = requests.get(url) # Делаем GET-запрос к ресурсу и результат ответа сохраняем в переменной response
page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup, указывая html-парсер

cities = str(page.find('table', class_='standard')) # берем города из нужной нам таблицы
cities = cities.split('title="')[3:] # отделяем ненужную информацию

million_cities = []
for i in cities:
    million_cities.append(i.split('">')[0])
    
million_cities = tuple(million_cities)
million_cities

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Красноярск',
 'Челябинск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [94]:
query_5_7 = f'''
-- Для компании «Яндекс» выводим список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах.
with x AS(
SELECT
    a.name,
    COUNT(*) cnt
FROM vacancies v
JOIN employers e ON v.employer_id = e.id
JOIN areas a ON v.area_id = a.id
WHERE e.name = 'Яндекс' AND a.name IN {million_cities}
GROUP BY a.id
)

SELECT
    x.name,
    x.cnt
FROM x

-- добавляем строку Total с общим количеством вакансий компании
UNION
SELECT
    'total',
    SUM(x.cnt)
FROM x
ORDER BY 2
'''

In [95]:
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_7720\3129374555.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7               Уфа   26.0
8            Самара   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            total  485.0

***

У большинства работадателей число вакансий равно 1, однако есть крупные компании, которые предоставляют большое количество вакансий, из которых сильно выделяется "Яндекс" с почти 2 тыс. вакансий. 
Похожая ситуация с регионами. У более чем 80% работадателей вакансии располагаются в одном регионе. Среди них так же сильно выделяется яндекс, у которого вакансии есть в 181 регионе. 

Более 80% работодателей, у которых указана сфера(ы) деятельности ограничиваются 1-3 сферами.

У 10% работодателей в качестве сферы деятельности указана 'Разработка программного обеспечения'

У 1/3 работадателей не указана сфера деятельности, некоторые модели включающие в себя этот параметр, могут быть неточными.

В городах-миллионниках компания Яндекс предоставляет в среднем по 30 вакансий, это примерно 1-3% от общего количества вакансий на город.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [239]:
query_6_1 = f'''
SELECT COUNT(*)
FROM vacancies
WHERE (lower(name) LIKE '%data%') OR (lower(name) LIKE '%данн%')
'''

In [240]:
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\485020324.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [14]:
query_6_2 = f'''
with ds as(SELECT *
FROM vacancies
WHERE
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name ILIKE '%ML%' AND name NOT ILIKE '%HTML%')
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%')
    
SELECT COUNT(*)
FROM ds
WHERE
    lower(name) LIKE '%junior%'
    OR experience = 'Нет опыта'
    OR employment = 'Стажировка'
'''

In [15]:
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_4952\3641650667.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [243]:
query_6_3 = f'''
with ds as(SELECT *
FROM vacancies
WHERE
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name ILIKE '%ML%' AND name NOT ILIKE '%HTML%')
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%')
    
SELECT COUNT(*)
FROM ds
WHERE
    lower(key_skills) LIKE '%sql%'
    OR lower(key_skills) LIKE '%postgres%'
'''

In [244]:
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\1159587760.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS. Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [245]:
query_6_4 = f'''
with ds as(SELECT *
FROM vacancies
WHERE
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name ILIKE '%ML%' AND name NOT ILIKE '%HTML%')
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%')
    
SELECT COUNT(*)
FROM ds
WHERE
    lower(key_skills) LIKE '%python%'
'''

In [246]:
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\972386973.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [247]:
query_6_5 = f'''
with ds as(
    SELECT *
    FROM vacancies
    WHERE
        (name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%')
        AND key_skills IS NOT NULL
)

-- Считаем количество разделителей.
-- Разделителей всегда меньше на 1, чем разделяемое. Поэтому +1. Это будет количеством ключевых навыков
SELECT round(avg(
    length(key_skills) - length(replace(key_skills, CHR(9), '')) +1
    ), 2)
FROM ds
'''

In [248]:
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_11140\390248827.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [16]:
query_6_6 = f'''
-- Запрос для DS? в которых указана з/п
with ds as(SELECT *
FROM vacancies
WHERE
    (name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%')
    
    AND
    (salary_from IS NOT NULL
    OR salary_to IS NOT NULL)
    )


SELECT 
    experience,
    -- считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
    ROUND(AVG(
    CASE WHEN (salary_from IS NOT NULL AND salary_to IS NOT NULL) THEN (salary_from + salary_to)/2
    WHEN (salary_from IS NOT NULL AND salary_to IS NULL) THEN salary_from ELSE salary_to
    END))
FROM ds
GROUP BY experience

'''

In [17]:
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_4952\3561045109.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience     round
0       От 3 до 6 лет  243115.0
1  От 1 года до 3 лет  139675.0
2           Нет опыта   74643.0

***

Из представленных вакансий 1771 из них имеет отношение к данным. Это примерно 3.5% от общего количества.

Меньше трети таких вакансий (536 вакансий) представлено на должность data scientist. Для удобства я буду называть их DSV (data science vacancy).
И 10% от DSV (51 вакансия) для начинающего дата-сайентиста.

В 2/5 DSV, как ключевые навыки указаны SQL или postgres. Это неплохой показатель того, что эти навыки достаточно востребованы.

Знание Python в требованиях работодателей к DS необходимо в 2/3 DSV. Так что хорошее знание языка хорошо увеличивает выбор вакансий для кандидата.

В среднем в DSV указывают 6.41 ключевых навыков. Владея небольшим количеством навыков, шансы на получение должности будут крайне малы. Чтобы иметь хорошие шансы на получение места в компании, лучше изучить какие ключевые навыки пользуются спросом и сделать выводы.

Как мы видим из выводов, зарплата для DSV без опыта работы - 75.000 руб. Набираясь опыта средняя з/п специалиста увеличивается примерно на 30.000-35.000 руб.

# Дополнительные исследования данных

1. Отличается ли з/п в вакансиях в которых указана "Полная занятость" от других вакансий

In [14]:
query_ad_1 = f'''
SELECT
    employment,
    -- считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
    ROUND(AVG(
    CASE WHEN (salary_from IS NOT NULL AND salary_to IS NOT NULL) THEN (salary_from + salary_to)/2
    WHEN (salary_from IS NOT NULL AND salary_to IS NULL) THEN salary_from ELSE salary_to
    END))
FROM vacancies
WHERE
    salary_from IS NOT NULL
    OR salary_to IS NOT NULL
GROUP BY employment
ORDER BY 2 DESC
'''

df = pd.read_sql_query(query_ad_1, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_7720\3342484730.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_ad_1, connection)


employment    round
0     Проектная работа  92086.0
1     Полная занятость  88240.0
2           Стажировка  37249.0
3  Частичная занятость  31566.0

2. Выведем среднюю з/п в двадцатке городов, по количеству вакансий и в отдельной строчке для всех остальных регионов.

In [63]:
query_ad_2 = f'''
-- Запрос для топ-20 городов по количеству вакансий
WITH x2 AS(
SELECT 
    a.name,
    COUNT(area_id) cnt,
    -- считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
    ROUND(AVG(
    CASE WHEN (salary_from IS NOT NULL AND salary_to IS NOT NULL) THEN (salary_from + salary_to)/2
    WHEN (salary_from IS NOT NULL AND salary_to IS NULL) THEN salary_from ELSE salary_to
    END)) avg_salary
FROM vacancies v
JOIN areas a ON a.id = v.area_id
WHERE
    salary_from IS NOT NULL
    OR salary_to IS NOT NULL
GROUP BY a.id
ORDER BY 2 DESC
LIMIT 20
),

-- Запрос для оставшихя городов
x1 AS(
SELECT 
    a.name,
    COUNT(area_id) cnt,
    ROUND(AVG(
    CASE WHEN (salary_from IS NOT NULL AND salary_to IS NOT NULL) THEN (salary_from + salary_to)/2
    WHEN (salary_from IS NOT NULL AND salary_to IS NULL) THEN salary_from ELSE salary_to
    END)) avg_salary
FROM vacancies v
JOIN areas a ON a.id = v.area_id
WHERE
    salary_from IS NOT NULL
    OR salary_to IS NOT NULL
GROUP BY a.id
ORDER BY 2 DESC
OFFSET 20
)




SELECT
    'top-20' areas,
    ROUND(AVG(avg_salary))
from x2
UNION ALL
SELECT
    'other',
    ROUND(AVG(avg_salary))
FROM x1

'''

df = pd.read_sql_query(query_ad_2, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_7720\3870917789.py:50: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_ad_2, connection)


areas    round
0  top-20  90888.0
1   other  63833.0

3. Выведем некоторые компании, количество вакансий в них и среднюю з/п и посмотрим есть ли зависимость з/п от кол-ва вакансий.

In [70]:
query_ad_3 = f'''
with x as(
SELECT 
    e.name,
    COUNT(v.id) cnt,
    -- считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
    ROUND(AVG(
    CASE WHEN (salary_from IS NOT NULL AND salary_to IS NOT NULL) THEN (salary_from + salary_to)/2
    WHEN (salary_from IS NOT NULL AND salary_to IS NULL) THEN salary_from ELSE salary_to
    END)) avg_salary
FROM vacancies v
JOIN employers e ON e.id = v.employer_id
WHERE
    salary_from IS NOT NULL
    OR salary_to IS NOT NULL
GROUP BY e.name
ORDER BY COUNT(v.id) DESC)

SELECT *
FROM x
LIMIT 20
'''

df = pd.read_sql_query(query_ad_3, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_7720\2370890531.py:23: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_ad_3, connection)


name   cnt  avg_salary
0                   Яндекс  1495     25884.0
1               Ростелеком   392     46980.0
2                   Тензор   155    171149.0
3               Спецремонт   151     70417.0
4               Первый Бит   132     92324.0
5   Поляков Денис Иванович   127     52638.0
6                      МТС    92     64521.0
7                 Andersen    86    269475.0
8                 1C-Рарус    77     81548.0
9                ООО ЕФИН     71    150000.0
10       DNS Головной офис    69    101152.0
11             ИК СИБИНТЕК    68     53434.0
12                   ANCOR    65     98208.0
13                   АЛНАС    56    136796.0
14          ЦИФРОВОЙ ВЫБОР    55    141455.0
15             Восток.Софт    55    111000.0
16                Homework    54     93333.0
17   Sumitec International    53    165000.0
18              ЭР-Телеком    53     50858.0
19            Почта России    51     35278.0

4. Выыведем топ-20 сфер деятельности по з/п в вакансиях, и посмотрим есть ли в них "Разработка програмного обеспечения".

In [91]:
query_ad_4 = f'''
with x as(
SELECT 
    i.name,
    -- считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
    ROUND(AVG(
    CASE WHEN (salary_from IS NOT NULL AND salary_to IS NOT NULL) THEN (salary_from + salary_to)/2
    WHEN (salary_from IS NOT NULL AND salary_to IS NULL) THEN salary_from ELSE salary_to
    END)) avg_salary
FROM vacancies v
JOIN employers e ON e.id = v.employer_id
JOIN employers_industries ea ON e.id = ea.employer_id
JOIN industries i ON ea.industry_id = i.id
WHERE
    salary_from IS NOT NULL
    OR salary_to IS NOT NULL
GROUP BY i.name
ORDER BY 2 DESC)

SELECT *
FROM x
LIMIT 20
'''


df = pd.read_sql_query(query_ad_4, connection)
df

C:\Users\SHOLFI\AppData\Local\Temp\ipykernel_7720\3455236080.py:25: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_ad_4, connection)


name  avg_salary
0           Агрохимия (продвижение, оптовая торговля)    381429.0
1                            Агрохимия (производство)    339875.0
2                                      Игорный бизнес    191905.0
3                   Корма для животных (производство)    168300.0
4                                          Зоомагазин    167613.0
5                                  Кадровые агентства    144530.0
6       Лакокрасочная продукция, сырье (производство)    142803.0
7   Бытовая химия, парфюмерия, косметика  (произво...    130120.0
8   Лакокрасочная продукция, сырье (продвижение, о...    128445.0
9   Дорожно-строительная техника, сельскохозяйстве...    122878.0
10  Дорожно-строительная техника, сельскохозяйстве...    119919.0
11  Оборудование для пищевой промышленности, упако...    119656.0
12  Финансово-кредитное посредничество (биржа, бро...    117714.0
13                Разработка программного обеспечения    117659.0
14  Управляющая, инвестиционная компания (управлен...    114297.0
15  Корма для животных (продвижение, оптовая торго...    113667.0
16                          Ветеринарная деятельность    113364.0
17                     Учреждение соц.помощи и защиты    113125.0
18                                Лизинговые компании    112306.0
19                       Пошив, ремонт одежды и обуви    104000.0

***

У вакансий с типами трудоустройства "Проектная работа" и "Полная занятость" средняя з/п в 2.5-3 раза больше чем у вакансий в уоторых указано "Стажировка" или "Частичная занятость". Что логично, так как в отличие от стажировки, для проектной работы требуются более компетентные сотрудники. А за полную занятость будут платить больше, нежели чем за частичную.

В городах, с большим количеством вакансий средняя з/п выше. На эту тему можно провести доп. исследования. При этом средняя з/п внутри компании не зависит от количества вакансий в ней.

Учитывая, что в нашей выборке 290 сфер деятельности, для которых УКАЗАНА з/п. Разработка программного обеспечения находится на 13 месте.

# Общий вывод по проекту

В выборке много вакансий с типом трудоустройства "Полная занятость" (93.6%). Однако не все компании требуют нахождения сотрудника в офисе полный рабочий день.
Так же в большей части ваканчий требуется опыт работы хотя бы 1 или больше лет. 

В городах, с большим количеством вакансий средняя з/п выше. Это может объясняться разными факторами: высокой конкуренцией, нехваткой специалистов или чем то ещё.

Помимо этого можно провести другие исследования:
* Выявить самые популярные ключевые навыки;
* Соотнесение з/п для DSV на позицию Junior к з/п начинающих должностей в других сферах;
* Сравнить российские города-миллионники с другими городами России по разным факторам.

In [96]:
connection.close()